airbyte-cdk/python/airbyte_cdk/sources/streams/abstract_stream.py

```diff
@@ -0,0 +1,213 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from abc import ABC, abstractmethod
+from functools import lru_cache
+from typing import Any, Iterable, List, Mapping, Optional, Tuple, Union
+
+import requests
+from airbyte_cdk.models import AirbyteMessage, AirbyteStream, SyncMode
+from airbyte_cdk.models import Type as MessageType
+from airbyte_cdk.sources.source import Source
+from airbyte_cdk.sources.utils import casing
+from airbyte_cdk.sources.utils.schema_helpers import InternalConfig
+from airbyte_cdk.sources.utils.slice_logger import SliceLogger
+from airbyte_cdk.sources.utils.transform import TransformConfig, TypeTransformer
+from airbyte_cdk.sources.utils.types import JsonType, StreamData
+
+
+class AbstractStream(ABC):
+    @abstractmethod
+    def read(
+        self,
+        cursor_field: Optional[List[str]],
+        logger: logging.Logger,
+        slice_logger: SliceLogger,
+        internal_config: InternalConfig = InternalConfig(),
+    ) -> Iterable[StreamData]:
+        """
+        Read a stream in full refresh mode
+        :param stream: The stream to read data from
+        :param cursor_field:
+        :param logger:
+        :param slice_logger:
+        :param internal_config:
+        :return: The stream's records
+        """
+
+    @property
+    def logger(self) -> logging.Logger:
```

Comment:
I saw `stream_slices`, `state_checkpoint_interval`, and `get_updated_state` as well but they were not in the code path yet so it made sense to exclude them. One note on these methods: they probably expose some internals of the stream for no real reasons. I would try to implement the concurrent CDK without exposing these. It feels natural to me to have the cursor do the checkpointing but I haven't thought about the details of this

Reply:
I would prefer waiting until we need the methods before adding them instead of committing right away to an interface so I left out all methods related to the state.

The code has change a bit, but I did add `stream_slices` to the facade interface. Good catch!